src/kvstore/raftex/RaftPart.cpp

```diff
@@ -382,22 +371,34 @@ folly::Future<AppendLogResult> RaftPart::appendLogAsync(ClusterID source,
                 retFuture = cachingPromise_.getSharedFuture();
                 break;
         }
-        if (replicatingLogs_) {
-            VLOG(2) << idStr_
-                    << "Another AppendLogs request is ongoing,"
-                       " just return";
-            return retFuture;
-        } else {
+
+        bool expected = false;
+        if (replicatingLogs_.compare_exchange_strong(expected, true)) {
             // We need to send logs to all followers
             VLOG(2) << idStr_ << "Preparing to send AppendLog request";
-            replicatingLogs_ = true;
             sendingPromise_ = std::move(cachingPromise_);
             cachingPromise_.reset();
             std::swap(swappedOutLogs, logs_);
-            firstId = lastLogId_ + 1;
+        } else {
+            VLOG(2) << idStr_
+                    << "Another AppendLogs request is ongoing,"
+                       " just return";
+            return retFuture;
         }
     }
 
+    {
+        std::lock_guard<std::mutex> g(raftLock_);
+        auto res = canAppendLogs();
+        if (res != AppendLogResult::SUCCEEDED) {
+            LOG(ERROR) << idStr_
+                       << "Cannot append logs, clean the buffer";
+            sendingPromise_.setValue(std::move(res));
```

Comment:
maybe no need move?

Reply:
Good point.  Actually, in this case move equals copy.  It is harmless.